homeassistant/helpers/restore_state.py

```diff
@@ -2,97 +2,145 @@
-from homeassistant.components.recorder import (
-    wait_connection_ready, DOMAIN as _RECORDER)
-import homeassistant.util.dt as dt_util
+DATA_RESTORE_STATE_TASK = 'restore_state_task'
 
-RECORDER_TIMEOUT = 10
-DATA_RESTORE_CACHE = 'restore_state_cache'
-_LOCK = 'restore_lock'
 _LOGGER = logging.getLogger(__name__)
 
+STORAGE_KEY = 'core.restore_state'
+STORAGE_VERSION = 1
+
+# How long between periodically saving the current states to disk
+STATE_DUMP_INTERVAL = timedelta(minutes=15)
+
+
+class RestoreStateData():
+    """Helper class for managing the helper saved data."""
+
+    @staticmethod
+    async def async_get_instance(hass: HomeAssistant) -> 'RestoreStateData':
+        """Get the singleton instance of this data helper."""
+        task = hass.data.get(DATA_RESTORE_STATE_TASK)
+
+        if task is None:
+            async def load_instance(hass: HomeAssistant) -> 'RestoreStateData':
+                """Set up the restore state helper."""
+                data = RestoreStateData(hass)
+
+                try:
+                    states = await data.store.async_load()
+                except HomeAssistantError as exc:
+                    _LOGGER.error("Error loading last states", exc_info=exc)
+                    states = None
+
+                if states is None:
+                    _LOGGER.debug('Not creating cache - no saved states found')
+                    data.last_states = {}
+                else:
+                    data.last_states = {
+                        state['entity_id']: State.from_dict(state)
+                        for state in states}
+                    _LOGGER.debug(
+                        'Created cache with %s', list(data.last_states))
+
+                if hass.state == CoreState.running:
+                    data.async_setup_dump()
+                else:
+                    hass.bus.async_listen_once(
+                        EVENT_HOMEASSISTANT_START, data.async_setup_dump)
+
+                return data
+
+            task = hass.data[DATA_RESTORE_STATE_TASK] = hass.async_create_task(
+                load_instance(hass))
+
+        return await task
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the restore state data class."""
+        self.hass = hass  # type: HomeAssistant
+        self.store = Store(hass, STORAGE_VERSION, STORAGE_KEY,
+                           encoder=JSONEncoder)  # type: Store
+        self.last_states = {}  # type: Dict[str, State]
+        self.entities = []  # type: List[RestoreEntity]
+
+    async def async_dump_states(self) -> None:
+        """Save the current state machine to storage."""
+        _LOGGER.debug("Dumping states")
+        # Entity ID set of registered restorable entities to dump
+        entity_ids = set(entity.entity_id for entity in self.entities)
+        try:
+            await self.store.async_save([
+                state.as_dict() for state in self.hass.states.async_all()
```

Comment:
By doing this we will lose any state of an entity that didn't show up in this run, but might show up in the future.

Wouldn't it make more sense to just dump the cache and during a dump update the cache with the latest?

Reply:
Dumping entities that don't show up in the current run is the behavior of the recorder implementation, although I'm not opposed to changing that. Would we then be keeping entities around indefinitely, or should we include an expiration date?

Reply:
Although, by relying on the cache, it will just grow forever, which is also bad.

Reply:
We could do an expiration, ie if an entity hasn't been seen for a week or so it is dropped.